### ISSUE TITLE
Feature request: Ability to specify local checked out branch with GitScm

### ISSUE THREAD

**Issue by eddgrant** [hmrc/jenkins-job-builders#16]
Hey folks,

When using GitScm Jenkins seems to default to checking out the local code in to a detached head state. This is usually fine, however when using the [gradle-release-plugin](https://github.com/researchgate/gradle-release) we need to be able to specify the checked out local branch name.

This can be achieved quite easily with a change to [GitHubScm](https://github.com/hmrc/jenkins-job-builders/blob/master/src/main/groovy/uk/gov/hmrc/jenkinsjobbuilders/domain/scm/GitHubScm.groovy) by doing the following:

* Add a new 'localBranch' instance variable (and the appropriate parameters in the static method(s) to set it)
* In `toDsl()` append the following DSL

```
    if (this.localBranch != null) {
      configure { node ->
        node / 'extensions' << 'hudson.plugins.git.extensions.impl.LocalBranch' {
          localBranch(this.localBranch)
        }
      }
    } 
```

I've made this change locally and it works well, was wondering if you'd be interested in having it in the library? Let me know if so and I'll raise a PR.

Cheers,

Edd

**Comment by agilestevesmith** [comment]
Hi @eddgrant 

Good to hear from you. Where is `hudson.plugins.git.extensions.impl.LocalBranch` sourced from? Is it part of a particular Jenkins plugin?

**Comment by eddgrant** [comment]
Hey @agilestevesmith 

Looks like it lives in the [git-plugin](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/hudson/plugins/git/extensions/impl/LocalBranch.java)

**Comment by agilestevesmith** [comment]
Agreed, so I don't think GitHubScm is the way to go. Looking at the API Viewer https://jenkinsci.github.io/job-dsl-plugin/ I think you are mixing and matching the GitHub and Git plugins. 

Do you want to just use git instead?

**Comment by eddgrant** [comment]
Ah yes I see what you mean. Any suggestions? Do you have a particular convention you try and follow when mapping DSL elements to builder classes?

**Comment by agilestevesmith** [comment]
The Single Responsibility Principle  ;)

**Comment by eddgrant** [comment]
I thought you might say that :-) So in practise here would you like to see an new `GitScm` class which performed the above DSL modification? If so how would you want to expose the functionality? There seem to be 2 initial possibilities to me:

1: Change `GitHubScm` and `GitHubComScm` to both extend (or compose with?) this class so that they can also expose its functionality e.g.

    jobBuilder.withScm(GitHubScm.gitHubScm("github.com", githubRepoName, "master", "ssh", "", null, "my-local-branch-name"))

2: Leave the new class as a standalone `SCM` class and require the the calling code to explicitly configure the individual SCM classes e.g.

    jobBuilder.withScm(GitScm.gitScm("github.com", githubRepoName, "master", "ssh", "", null, "my-local-branch-name"))
    jobBuilder.withScm(GitHubScm.gitHubScm("github.com", githubRepoName, "master", "ssh", "", null))
    
My view is that [2] is somewhat confusing as it feels like 2 SCMs are being configured when in fact only 1 is. Furthermore `JobBuilder` would require modification in order to support this approach as `withScm` only currently supports a single SCM. In fact it might not work at all given that both SCMs instances would be configuring different properties on the same Git repo.

[1] feels much more elegant to me but it would couple the different SCM classes either by composition or inheritance. I also don't know whether the Closure returned by `GitScm` `toDsl()` could be further modified by `GitHubScm` etc. Should be easy to write a test to prove this though.

If you can provide some more insight r.e. your preferred approach then I'll put something together.

**Comment by agilestevesmith** [comment]
Supporting a second SCM seems wrong. but I don't see how the two SCM plugins are woven together by Jenkins. Could you send a screenshot of what it looks like in Jenkins please?
